**Alumno:** César Emilio García Ávalos

**Actividad:** M5_AI1_Modelo lineal Gaussiano

In [ ]:
library(readr)
library(tidyr)
library(dplyr)
install.packages("car")
library(car)
install.packages("broom")
library(broom)
data(Salaries)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [ ]:
summary(Salaries)

        rank     discipline yrs.since.phd    yrs.service        sex     
 AsstProf : 67   A:181      Min.   : 1.00   Min.   : 0.00   Female: 39  
 AssocProf: 64   B:216      1st Qu.:12.00   1st Qu.: 7.00   Male  :358  
 Prof     :266              Median :21.00   Median :16.00               
                            Mean   :22.31   Mean   :17.61               
                            3rd Qu.:32.00   3rd Qu.:27.00               
                            Max.   :56.00   Max.   :60.00               
     salary      
 Min.   : 57800  
 1st Qu.: 91000  
 Median :107300  
 Mean   :113706  
 3rd Qu.:134185  
 Max.   :231545  

In [ ]:
str(Salaries)

'data.frame':	397 obs. of  8 variables:
 $ rank         : Factor w/ 3 levels "AsstProf","AssocProf",..: 3 3 1 3 3 2 3 3 3 3 ...
 $ discipline   : Factor w/ 2 levels "A","B": 2 2 2 2 2 2 2 2 2 2 ...
 $ yrs.since.phd: int  19 20 4 45 40 6 30 45 21 18 ...
 $ yrs.service  : int  18 16 3 39 41 6 23 45 20 18 ...
 $ sex          : Factor w/ 2 levels "Female","Male": 2 2 2 2 2 2 2 2 2 1 ...
 $ salary       : int  139750 173200 79750 115000 141500 97000 175000 147765 119250 129000 ...
 $ salary_miles : num  139.8 173.2 79.8 115 141.5 ...
 $ log_salary   : num  11.8 12.1 11.3 11.7 11.9 ...


Se observa que no hay datos nulos, se procede a trabajar en el dataframe.

In [ ]:
# Realizar la regresión lineal
modelo1 <- lm(salary ~ yrs.service + yrs.since.phd, data = Salaries)

# Resumen del modelo
tidy(modelo1)
summary(modelo1)

term          estimate   std.error statistic p.value      
1 (Intercept)   89912.1845 2843.5599 31.619586 3.810703e-110
2 yrs.service    -629.1014  254.4694 -2.472208  1.384961e-02
3 yrs.since.phd  1562.8889  256.8196  6.085551  2.753560e-09


Call:
lm(formula = salary ~ yrs.service + yrs.since.phd, data = Salaries)

Residuals:
   Min     1Q Median     3Q    Max 
-79735 -19823  -2617  15149 106149 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)    89912.2     2843.6  31.620  < 2e-16 ***
yrs.service     -629.1      254.5  -2.472   0.0138 *  
yrs.since.phd   1562.9      256.8   6.086 2.75e-09 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 27360 on 394 degrees of freedom
Multiple R-squared:  0.1883,	Adjusted R-squared:  0.1842 
F-statistic: 45.71 on 2 and 394 DF,  p-value: < 2.2e-16


In [ ]:
# Obtener los residuos del modelo
residuos1 <- residuals(modelo)

# Calcular la suma del cuadrado de los residuos
RSS1 <- sum(residuos1^2)

# Mostrar la suma del cuadrado de los residuos
print(RSS1)

[1] 294874679053


*Dirección inesperada del coeficiente de yrs.service: El signo negativo del coeficiente asociado a la variable yrs.service sugiere que, contrariamente a lo esperado, un aumento en un año de antigüedad se relaciona con una disminución en el salario en 629 dólares. Esto va en contra de la intuición común, que sugeriría que a medida que uno gana más experiencia, debería recibir una compensación salarial más alta. Este resultado podría indicar que hay otros factores no considerados en el modelo que están influyendo en esta relación.

*Bajo coeficiente de determinación (R²): El coeficiente de determinación (R²) del modelo es bajo, lo que significa que solo aproximadamente el 18.8% de la variabilidad en el salario se explica por las variables predictoras incluidas en el modelo. Esto sugiere que el modelo no captura adecuadamente la variabilidad en el salario, y hay otras variables importantes que no están siendo consideradas.

*Predictores estadísticamente significativos: A pesar del bajo R², los predictores (años de servicio y años desde el doctorado) son estadísticamente significativos (Pvalue es menor al nivel de significancia del 0.05 para ambas variables). Esto sugiere que, aunque el modelo no explica mucha variabilidad en el salario, las variables predictoras aún están asociadas de manera significativa con el salario.

In [ ]:
# Realizar la regresión lineal incluyendo el género
modelo_con_genero <- lm(salary ~ yrs.service + yrs.since.phd + sex, data = Salaries)

# Resumen del modelo
tidy(modelo_con_genero)
summary(modelo_con_genero)

term          estimate   std.error statistic p.value     
1 (Intercept)   82875.9118 4800.6302 17.263548 4.194022e-50
2 yrs.service    -649.7608  253.9850 -2.558264 1.089370e-02
3 yrs.since.phd  1552.7571  256.1343  6.062276 3.150384e-09
4 sexMale        8457.0650 4656.1372  1.816326 7.008205e-02


Call:
lm(formula = salary ~ yrs.service + yrs.since.phd + sex, data = Salaries)

Residuals:
   Min     1Q Median     3Q    Max 
-79586 -19564  -3018  15071 105898 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)    82875.9     4800.6  17.264  < 2e-16 ***
yrs.service     -649.8      254.0  -2.558   0.0109 *  
yrs.since.phd   1552.8      256.1   6.062 3.15e-09 ***
sexMale         8457.1     4656.1   1.816   0.0701 .  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 27280 on 393 degrees of freedom
Multiple R-squared:  0.1951,	Adjusted R-squared:  0.189 
F-statistic: 31.75 on 3 and 393 DF,  p-value: < 2.2e-16


In [ ]:
# Obtener los residuos del modelo
residuos2 <- residuals(modelo_con_genero)

# Calcular la suma del cuadrado de los residuos
RSS2 <- sum(residuos2^2)

# Mostrar la suma del cuadrado de los residuos
print(RSS2)

[1] 2.9242e+11


In [ ]:
if (RSS1 > RSS2) {
  print("El modelo con la variable 'genero' tiene una menor suma de residuos al cuadrado")
} else {
  print("El modelo 1 tiene una menor suma de residuos al cuadrado")
}

[1] "El modelo con la variable 'genero' tiene una menor suma de residuos al cuadrado"


El coeficiente de determinación del modelo 1 es 18.42% y el coeficiente de determinación del modelo 2 es 18.9%, existe un incremento sin embargo es pequeño, notar que la variable género no es estadísticamente significativa en cuanto al impacto en el salario de un profesor (Pvalue = 0.07 mayor que el nivel de significancia de 0.05)

Al tomar en cuenta el modelo1, se identifica que manteniendo constante el número de años desde el doctorado, el salario disminuye en $629.1 por cada año adicional de servicio.

In [ ]:
# Dividir el salario por mil para expresarlo en miles
Salaries$salary_miles <- Salaries$salary / 1000

# Realizar la regresión lineal con el salario expresado en miles
modelo_salario_miles <- lm(salary_miles ~ yrs.service + yrs.since.phd, data = Salaries)

# Resumen del modelo
tidy(modelo_salario_miles)
summary(modelo_salario_miles)

term          estimate   std.error statistic p.value      
1 (Intercept)   89.9121845 2.8435599 31.619586 3.810703e-110
2 yrs.service   -0.6291014 0.2544694 -2.472208  1.384961e-02
3 yrs.since.phd  1.5628889 0.2568196  6.085551  2.753560e-09


Call:
lm(formula = salary_miles ~ yrs.service + yrs.since.phd, data = Salaries)

Residuals:
    Min      1Q  Median      3Q     Max 
-79.735 -19.823  -2.617  15.149 106.149 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)    89.9122     2.8436  31.620  < 2e-16 ***
yrs.service    -0.6291     0.2545  -2.472   0.0138 *  
yrs.since.phd   1.5629     0.2568   6.086 2.75e-09 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 27.36 on 394 degrees of freedom
Multiple R-squared:  0.1883,	Adjusted R-squared:  0.1842 
F-statistic: 45.71 on 2 and 394 DF,  p-value: < 2.2e-16


Al observar los coeficientes de ambos modelos (modelo1 y modelo_salario_miles), se identifica que los betas son reescalados también al reescalar la variable dependiente. Se observa que los valores de los mismos, fueron también divididos entre 1000.

In [ ]:
# Aplicar el logaritmo neperiano al salario
Salaries$log_salary <- log(Salaries$salary)

# Realizar la regresión lineal con el logaritmo neperiano del salario
modelo_log_salario <- lm(log_salary ~ yrs.service + yrs.since.phd, data = Salaries)

# Resumen del modelo
tidy(modelo_log_salario)
summary(modelo_log_salario)

term          estimate     std.error   statistic  p.value     
1 (Intercept)   11.400345338 0.024351888 468.150363 0.000000e+00
2 yrs.service   -0.005316121 0.002179244  -2.439434 1.515066e-02
3 yrs.since.phd  0.013471319 0.002199371   6.125079 2.196393e-09


Call:
lm(formula = log_salary ~ yrs.service + yrs.since.phd, data = Salaries)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.85152 -0.17076 -0.00342  0.15606  0.64229 

Coefficients:
               Estimate Std. Error t value Pr(>|t|)    
(Intercept)   11.400345   0.024352 468.150  < 2e-16 ***
yrs.service   -0.005316   0.002179  -2.439   0.0152 *  
yrs.since.phd  0.013471   0.002199   6.125  2.2e-09 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.2343 on 394 degrees of freedom
Multiple R-squared:  0.1932,	Adjusted R-squared:  0.1892 
F-statistic: 47.19 on 2 and 394 DF,  p-value: < 2.2e-16


Al aplicar el logaritmo neperiano al salario no se identifica un cambio en los signos de las betas siendo el intercepto y el beta de la variable yrs.since.phd, positivos en ambos casos, mientras que el beta de la variable yrs.service es negativo en ambos modelos.

En el nuevo modelo donde el salario ha sido transformado utilizando el logaritmo neperiano (logaritmo natural), los coeficientes asociados a las variables predictoras representan el cambio porcentual en el salario por cada unidad de cambio en la variable predictora, manteniendo todas las demás variables constantes. Esto indica que, manteniendo constante el número de años desde el doctorado, el salario disminuye en aproximadamente un 0.53% por cada año adicional de servicio.

In [ ]:
# Modelo de regresión lineal del salario
modelo_propuesto <- lm(salary ~ yrs.service + yrs.since.phd + discipline + rank, data = Salaries)

# Resumen del modelo
tidy(modelo_propuesto)
summary(modelo_propuesto)

term          estimate   std.error statistic p.value     
1 (Intercept)   69869.0110 3332.1448 20.968180 5.828862e-66
2 yrs.service    -476.7179  211.8312 -2.250461 2.497485e-02
3 yrs.since.phd   534.6313  241.1593  2.216922 2.720308e-02
4 disciplineB   14505.1514 2343.4181  6.189741 1.523745e-09
5 rankAssocProf 12831.5375 4147.6685  3.093675 2.118982e-03
6 rankProf      45287.6890 4236.6534 10.689496 1.436047e-23


Call:
lm(formula = salary ~ yrs.service + yrs.since.phd + discipline + 
    rank, data = Salaries)

Residuals:
   Min     1Q Median     3Q    Max 
-65244 -13498  -1455   9638  99682 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)    69869.0     3332.1  20.968  < 2e-16 ***
yrs.service     -476.7      211.8  -2.250  0.02497 *  
yrs.since.phd    534.6      241.2   2.217  0.02720 *  
disciplineB    14505.2     2343.4   6.190 1.52e-09 ***
rankAssocProf  12831.5     4147.7   3.094  0.00212 ** 
rankProf       45287.7     4236.7  10.689  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 22550 on 391 degrees of freedom
Multiple R-squared:  0.4525,	Adjusted R-squared:  0.4455 
F-statistic: 64.64 on 5 and 391 DF,  p-value: < 2.2e-16


Para el modelo propuesto no se ha considerado únicamente la variable "género", el resto de las variables analizadas tienen un valor P debajo del nivel de significancia de 0.05 lo que indica que son estadísticamente significativas para explicar la variación en el salario de los docentes. El coeficiente de determinación para el modelo propuesto es de 44.55% mejor que el resto de los modelos analizados.